libkbfs/file_data.go

```diff
@@ -1523,37 +1835,84 @@ func (fd *fileData) deepCopy(ctx context.Context, codec kbfscodec.Codec,
 // undupChildrenInCopy takes a top block that's been copied via
 // deepCopy(), and un-deduplicates all leaf children of the block.  It
 // adds all child blocks to the provided `bps`, including both the
-// ones that were deduplicated and the ones that weren't (TODO).  It
-// returns the BlockInfos for all children, including ones that
-// weren't de-duplicated.
+// ones that were deduplicated and the ones that weren't.  It returns
+// the BlockInfos for all children.
 func (fd *fileData) undupChildrenInCopy(ctx context.Context,
 	bcache BlockCache, bops BlockOps, bps *blockPutState,
 	topBlock *FileBlock) ([]BlockInfo, error) {
 	if !topBlock.IsInd {
 		return nil, nil
 	}
 
-	// TODO: support multiple levels of indirection.
-	// TODO: parallelize
-	blockInfos := make([]BlockInfo, len(topBlock.IPtrs))
-	for i, iptr := range topBlock.IPtrs {
-		if iptr.RefNonce == kbfsblock.ZeroRefNonce {
-			// No need to un-dup mid-level indirect blocks.  TODO:
-			// fetch the blocks and add them to the bps (only needed
-			// for multiple levels of indirection).
-			blockInfos = append(blockInfos, iptr.BlockInfo)
-			continue
-		}
-		childBlock, _, err := fd.getter(ctx, fd.kmd, iptr.BlockPointer,
-			fd.file, blockRead)
+	// For indirect files, get all the paths to leaf blocks.  Note
+	// that because topBlock is a deepCopy, all of the blocks are also
+	// deepCopys and thus are modifiable.
+	pfr, err := fd.getIndirectBlocksForOffsetRange(ctx, topBlock, 0, -1)
+	if err != nil {
+		return nil, err
+	}
+	if len(pfr) == 0 {
+		return nil, fmt.Errorf(
+			"Indirect file %v had no indirect blocks", fd.rootBlockPointer())
+	}
+
+	// Append the leaf block to each path, since readyHelper expects
+	// it.
+	for i, path := range pfr {
+		leafPtr := path[len(path)-1].childIPtr().BlockPointer
+		leafBlock, _, err := fd.getter(
+			ctx, fd.kmd, leafPtr, fd.file, blockWrite)
```

Comment:
If we try to let the OS do it, there's no guarantee it will do it from kbfsfuse, i.e. we will start greatly affecting the performance of other apps.  I think people don't look kindly on processes that don't keep their own memory under reasonable limits.

The "explosion" happens in this case:

1. Device A appends 1 bytes to a large file F.
2. Device B (with journaling enabled) also appends 1 byte to large file F at the same time.
3. Device A's write makes it to the server first.
4. Device B tries to write to the server, but hits a conflict, and so makes a local branch.
5. Device B sees that the merged branch also has a write to file F.  To resolve this conflict, device B needs to make a copy of file F with a conflict prefix.
6. Because journaling doesn't support adding references to existing blocks, CR has to "undup" the blocks, meaning it has to download each block, and ready it again, generating a new block key and hence a new block ID.
7. It keeps all of those block copies in memory at once, until they are written out to the journal and CR finishes.

So yes, it probably depends on network I/O (unless those blocks are already cache locally).  But a slow network just means the blocks that are downloaded will stay in memory longer.